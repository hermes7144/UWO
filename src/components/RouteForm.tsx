import { useState, useEffect } from 'react';
import useRoute from '../Hooks/useRoute';
import Button from './ui/Button';
import { useLocation, useNavigate } from 'react-router-dom';
import { useUWORouteContext } from '../context/UWORouteContext';
type RouteType = {
  id?: string;
  title?: string;
  description?: string;
  major_goods?: string;
  major_chk?: boolean;
};

export default function RouteForm() {
  const { citys } = useUWORouteContext();
  const { addOrUpdateItem } = useRoute();
  const { state } = useLocation();
  const [route, setRoute] = useState<RouteType>({});
  const [success, setSuccess] = useState('');
  const navigate = useNavigate();

  useEffect(() => {
    state && setRoute(state.route);
  }, [state]);

  const handleChange = (e) => {
    const { name, value, checked } = e.target;
    if (name === 'major_chk') {
      const major_chk = checked ? true : false;
      setRoute((route) => ({ ...route, major_chk }));
    } else {
      setRoute((route) => ({ ...route, [name]: value }));
    }
  };

  const handleSubmit = (e) => {
    e.preventDefault();
    setRoute((route) => ({ ...route, major_goods: String(route.major_goods).replaceAll(' ', '') }));
    addOrUpdateItem.mutate(
      { citys, route },
      {
        onSuccess: () => {
          setSuccess('성공적으로 경로가 추가되었습니다.');
          setTimeout(() => {
            setSuccess(null);
            navigate('/routes');
          }, 1000);
        },
      }
    );
  };
  return (
    <form className='flex flex-col px-12 font-semibold' onSubmit={handleSubmit}>
      {success && <p className='my-2'>✔️{success}</p>}
      <input name='title' value={route.title ?? ''} placeholder='제목' required onChange={handleChange} />
      <textarea rows={5} className='resize-none border border-gray-300 px-2 py-2 focus:outline-none my-2' name='description' value={route.description ?? ''} placeholder=' 설명' onChange={handleChange}></textarea>
      <div className='flex justify-between items-center'>
        <input className='w-11/12' name='major_goods' value={route.major_goods || ''} placeholder='주요교역품' onChange={handleChange} />
        <input className='w-4 h-4 mx-3' name='major_chk' type='checkbox' checked={route.major_chk || false} onChange={handleChange} />
      </div>
      <Button text={'Save'}></Button>
    </form>
  );
}
