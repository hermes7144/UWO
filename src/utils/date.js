// import { format, register } from 'timeago.js';
// import koLocale from 'timeago.js/lib/lang/ko';

// register('ko', koLocale);

// export function formatAgo(date, lang = 'en_US') {
//   return format(date, lang);
// }

export function formatDate(rawDate) {
  const date = new Date(rawDate);
  const year = date.getFullYear().toString().slice(-2); // 연도에서 뒤의 두 자리만 추출
  const month = ('0' + (date.getMonth() + 1)).slice(-2); // 월을 두 자리 숫자로 변환
  const day = ('0' + date.getDate()).slice(-2); // 일을 두 자리 숫자로 변환
  return `${year}/${month}/${day}`;
}