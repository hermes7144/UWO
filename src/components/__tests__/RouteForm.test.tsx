import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { Route } from 'react-router-dom';
import { withAllContexts, withRouter } from '../../tests/utils';
import RouteForm from '../RouteForm';

describe('RouteForm', () => {
  const fakeUseRoute = jest.fn();
  it('renders correctly', () => {
    const { asFragment } = renderRouteForm();
    expect(asFragment()).toMatchSnapshot();
  });

  beforeEach(() => {
    fakeUseRoute.mockImplementation(() => {
      return { addOrUpdateItem: { mutate: () => true } };
    });
  });

  it('submit route', async () => {
    renderRouteForm();

    const searchKeyword = '제목입니다';
    const title = screen.getByPlaceholderText('제목');
    await userEvent.type(title, searchKeyword);
    expect(screen.getByDisplayValue('제목입니다')).toBeInTheDocument();

    const checkBox = screen.getByRole('checkbox');
    await userEvent.click(checkBox);
    expect(checkBox).toBeChecked();

    const submit = screen.getByRole('button');
    await userEvent.click(submit);
  });

  function renderRouteForm() {
    return render(withAllContexts(withRouter(<Route path='/' element={<RouteForm />} />), fakeUseRoute));
  }
});
