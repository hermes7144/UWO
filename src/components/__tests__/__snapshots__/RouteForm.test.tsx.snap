// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RouteForm renders correctly 1`] = `
<DocumentFragment>
  <form
    class="flex flex-col px-1 font-semibold"
  >
    <input
      maxlength="50"
      name="title"
      placeholder="제목"
      required=""
      value=""
    />
    <textarea
      class="resize-none border border-gray-300 px-2 py-2 focus:outline-none my-2"
      name="description"
      placeholder=" 설명"
      rows="5"
    />
    <div
      class="flex justify-between items-center"
    >
      <input
        class="w-11/12"
        name="major_goods"
        placeholder="주요교역품"
        value=""
      />
      <input
        class="w-4 h-4 mx-3"
        name="major_chk"
        type="checkbox"
      />
    </div>
    <div
      class="flex items-center"
    >
      <label
        class="text-brand font-bold"
        for="select"
      >
        국가:
      </label>
      <select
        class="p-1 m-1 flex-1 border-2 outline-none"
      >
        <option>
          공통
        </option>
        <option>
          포트투갈
        </option>
        <option>
          에스파냐
        </option>
        <option>
          잉글랜드
        </option>
        <option>
          네덜란드
        </option>
        <option>
          오스만
        </option>
        <option>
          명
        </option>
        <option>
          조선
        </option>
        <option>
          일본
        </option>
      </select>
      <label
        class="text-brand font-bold"
        for="select"
      >
        지역:
      </label>
      <select
        class="p-1 m-1 flex-1 border-2 outline-none"
      >
        <option>
          공통
        </option>
        <option>
          동지중해
        </option>
        <option>
          서지중해
        </option>
        <option>
          북해
        </option>
        <option>
          카리브해
        </option>
        <option>
          서아프리카
        </option>
        <option>
          동아프리카
        </option>
        <option>
          아랍
        </option>
        <option>
          서인도
        </option>
        <option>
          동인도
        </option>
        <option>
          서남아
        </option>
        <option>
          동남아
        </option>
        <option>
          동아시아
        </option>
        <option>
          기타
        </option>
      </select>
      <label
        class="text-brand font-bold"
        for="select"
      >
        월:
      </label>
      <select
        class="p-1 m-1 flex-1 border-2  outline-none"
      >
        <option>
          1
        </option>
        <option>
          2
        </option>
        <option>
          3
        </option>
        <option>
          4
        </option>
        <option>
          5
        </option>
        <option>
          6
        </option>
        <option>
          7
        </option>
        <option>
          8
        </option>
        <option>
          9
        </option>
        <option>
          10
        </option>
        <option>
          11
        </option>
        <option>
          12
        </option>
      </select>
      ~
      <select
        class="p-1 m-1 flex-1 border-2 outline-none"
      >
        <option>
          1
        </option>
        <option>
          2
        </option>
        <option>
          3
        </option>
        <option>
          4
        </option>
        <option>
          5
        </option>
        <option>
          6
        </option>
        <option>
          7
        </option>
        <option>
          8
        </option>
        <option>
          9
        </option>
        <option>
          10
        </option>
        <option>
          11
        </option>
        <option>
          12
        </option>
      </select>
    </div>
    <button
      class="bg-brand text-white
 py-2 px-4 rounded-sm hover:brightness-110"
    >
      Save
    </button>
  </form>
</DocumentFragment>
`;
